File: cicd_dagger_contract/python/test.example.py
"""Test module for Goserv"""

import json
import dagger
from dagger import dag, function, object_type
from typing import Optional


@object_type
class Test:
    @function
    async def unit_test(
        self,
        source: dagger.Directory,
        build_artifact: Optional[dagger.File] = None,
    ) -> str:
        """UnitTest runs the goserv container and executes unit tests against it
        
        Args:
            source: Source directory containing the project
            build_artifact: Build output from the Build function (if not provided, will build from source)
        
        Returns:
            Test output string
        """
        # Print message
        output = await (
            dag.container()
            .from_("alpine:latest")
            .with_exec(["echo", "this is the UnitTest function"])
            .stdout()
        )

        return output

    @function
    async def integration_test(
        self,
        source: dagger.Directory,
        deployment_context: Optional[dagger.File] = None,
        validation_context: Optional[dagger.File] = None,
    ) -> str:
        """IntegrationTest runs integration tests against a deployed goserv instance
        
        Args:
            source: Source directory containing the project
            deployment_context: Deployment context from Deploy function
            validation_context: Validation context from Validate function
        
        Returns:
            Test output string
        """
        # Extract endpoint from deployment context if provided
        target_url = None
        if deployment_context:
            context_content = await deployment_context.contents()
            context = json.loads(context_content)
            target_url = context.get("endpoint")

        # Check validation status if provided
        if validation_context:
            val_content = await validation_context.contents()
            val_context = json.loads(val_content)
            if val_context.get("status") != "healthy":
                return "Skipping tests: deployment validation failed"

        # Run integration tests against target_url
        output = await (
            dag.container()
            .from_("alpine:latest")
            .with_exec(["echo", f"Running integration tests against {target_url}"])
            .stdout()
        )
        output = await (
            dag.container()
            .from_("alpine:latest")
            .with_exec(["echo", "this is the IntegrationTest function"])
            .stdout()
        )

        return output

File: cicd_dagger_contract/typescript/test.example.ts
/**
 * Test module for Goserv
 */

import { dag, Container, Directory, File, Secret, object, func } from "@dagger.io/dagger"

@object()
export class Test {
  /**
   * UnitTest runs the goserv container and executes unit tests against it
   *
   * @param source Source directory containing the project
   * @param buildArtifact Build output from the Build function (if not provided, will build from source)
   * @returns Test output string
   */
  @func()
  async unitTest(
    source: Directory,
    buildArtifact?: File
  ): Promise<string> {
    // Print message
    const output = await dag
      .container()
      .from("alpine:latest")
      .withExec(["echo", "this is the UnitTest function"])
      .stdout()

    return output
  }

  /**
   * IntegrationTest runs integration tests against a deployed instance
   *
   * @param source Source directory containing the project
   * @param kubeconfig Kubernetes config file content
   * @param awsconfig AWS configuration file content
   * @param deploymentContext Deployment context from Deploy function
   * @param validationContext Validation context from Validate function
   * @returns String containing test results
   */
  @func()
  async integrationTest(
    source: Directory,
    kubeconfig: Secret,
    awsconfig?: Secret,
    deploymentContext?: File,
    validationContext?: File
  ): Promise<string> {
    // Extract endpoint from deployment context if provided
    let targetUrl: string | undefined
    if (deploymentContext) {
      const contextContent = await deploymentContext.contents()
      const context = JSON.parse(contextContent)
      targetUrl = context.endpoint
    }

    // Check validation status if provided
    if (validationContext) {
      const valContent = await validationContext.contents()
      const valContext = JSON.parse(valContent)
      if (valContext.status !== "healthy") {
        return "Skipping tests: deployment validation failed"
      }
    }

    // Run integration tests against targetUrl
    const output = await dag
      .container()
      .from("alpine:latest")
      .withExec(["echo", `Running integration tests against ${targetUrl}`])
      .stdout()

    return output
  }
}
